MSI-2043: disabled source selection when disabled manage stock

VitaliyBoyko opened the pull request:
### Description
Disabled source selection algorithm when managing stock is disabled.

### Fixed Issues (if relevant)
1. magento-engcom/msi#2043: Cannot create shipment: Source item not found by source code: default and SKU: [my-sku custom option] 

### Manual testing scenarios
#### 1
1. Create multiple sources and stocks
2. Create a product and assign it to the custom sources.
3. Disable stock managing Stores > Configuration > Catalog > Inventory > Product Stock Options > Manage Stock
4. Create an order with the product
5. Create and submit a shipment
SSA page shouldn't be shown.

#### 2
1. Create multiple sources and stocks
2. Create a product and assign it to the custom sources.
3. Disable stock managing for Product via Advanced Inventory
4. Create an order with the product
5. Create and submit a shipment
SSA page shouldn't be shown.

#### 3
1. Create multiple sources and stocks
2. Create 2 products and assign them to the custom sources.
3. Disable stock managing for 1 Product via Advanced Inventory
4. Create an order with both products
5. SSA page should be shown.
The issue with select sources for this scenario fixed in the scope of #2012.

Commit:
MSI-2043: disabled source selection when disabled manage stock

Commit:
MSI-2043: fixed static tests

adarshkhatri commented:
I think we need to check this:

```
        if (!$this->stockConfiguration->getManageStock()) {
            return;
        }
```

for each product, instead of entire site.

VitaliyBoyko commented:
Seems like we have to implement service which will check if the order has at least one stock manageable item. And show SSA according to result. And on SSA level we have to add the check whether item stock manageable or not. @maghamed I'm right?

Commit:
MSI-2043: check ech order item for manageability

maghamed commented:
@VitaliyBoyko not sure that we need to introduce a dedicated service for these purposes.
Just okay to introduce dedicated class which would use InventoryConfiguration inside

Commit:
MSI-2043: fixed product types

adarshkhatri commented:
It looks like there is check in place:

            if (!$stockItemConfiguration->isManageStock()) {
                //We don't need to Manage Stock
                continue;
            }

https://github.com/magento-engcom/msi/blob/2.3-develop/app/code/Magento/InventorySourceDeductionApi/Model/SourceDeductionService.php#L76

I think solution would be to find the list by `product id` instead of sku. I see that when creating order `Magento` adds `Custom Option` to the configurable item.

For example: 
Config item's sku: Config Item with Custom Option
Child Item: Config Item with Custom Option-Black-First

But when order created:
![db](https://user-images.githubusercontent.com/8390252/54662973-eb068880-4b33-11e9-83a7-fdfb5f0d0fc0.png)

adarshkhatri commented:
It looks like there is check in place:

            if (!$stockItemConfiguration->isManageStock()) {
                //We don't need to Manage Stock
                continue;
            }

https://github.com/magento-engcom/msi/blob/2.3-develop/app/code/Magento/InventorySourceDeductionApi/Model/SourceDeductionService.php#L76

I think solution would be to find the list by `product id` instead of sku. I see that when creating order `Magento` adds `Custom Option` to the configurable item, when it has custom option

**For example:** 
Config item's sku: Config Item with Custom Option
Child Item: Config Item with Custom Option-Black-First

But when order created:
![db](https://user-images.githubusercontent.com/8390252/54662973-eb068880-4b33-11e9-83a7-fdfb5f0d0fc0.png)

Because of this, system will never find the product.

       //namespace Magento\InventorySourceDeductionApi\Model\GetSourceItemBySourceCodeAndSku

        $searchCriteria = $this->searchCriteriaBuilder
            ->addFilter(SourceItemInterface::SOURCE_CODE, $sourceCode) //default
            ->addFilter(SourceItemInterface::SKU, $sku) //Config Item with Custom Option-Black-First Custom Option
            ->create();
        $sourceItemsResult = $this->sourceItemRepository->getList($searchCriteria);

**I think, finding the source item by `product id` will always find the product if it exist.**

ishakhsuvarov committed:
magento-engcom#2049: Merge branch '2.3-develop' of github.com:magento-engcom/msi into MSI-2043-disable-source-selection-when-disabled-manage-stock

Conflicts:
	app/code/Magento/InventoryShippingAdminUi/Observer/NewShipmentLoadBefore.php

ishakhsuvarov committed:
magento-engcom/msi#2049: disabled source selection when disabled manage stock
 - Resolved unwanted dependencies

malesh-v commented:
I've checked the first case, I click Ship and SSA page isn't showing(as expected) but if a click **Back** SSA page loads.

Commit:
MSI-2043: fixed back button on shipping step

Commit:
MSI-2043: fixed back button url

Commit:
MSI-2043: fixed back button on shipping step test

ishakhsuvarov commented on app/code/Magento/InventoryShippingAdminUi/Plugin/Sales/Block/Shipment/BackButtonUrlOnNewShipmentPagePlugin.php:
```diff
@@ -7,42 +7,59 @@
+        IsOrderSourceManageable $isOrderSourceManageable = null
     ) {
         $this->isWebsiteInMultiSourceMode = $isWebsiteInMultiSourceMode;
+        $this->isOrderSourceManageable = $isOrderSourceManageable ??
```
Plugins are not covered with Backward Compatibility policy, constructor may be updated without additional checks

ishakhsuvarov commented on app/code/Magento/InventoryShippingAdminUi/Plugin/Sales/Block/Shipment/BackButtonUrlOnNewShipmentPagePlugin.php:
```diff
@@ -7,42 +7,59 @@
 use Magento\InventoryShippingAdminUi\Model\IsWebsiteInMultiSourceMode;
 
+/**
+ * Class BackButtonUrlOnNewShipmentPagePlugin
```
This comment does not cover the purpose of the Plugin

Commit:
MSI-2043: removed object manager from plugin and fixed class annotation

VitaliyBoyko commented:
Hi @ishakhsuvarov 
I've added the mentioned changes.
Thank you!

ghost commented:
Hi @VitaliyBoyko, thank you for your contribution! </br>Please, complete [Contribution Survey](https://opensource.magento.com/survey/contribution/magento-engcom+msi+2049), it will take less than a minute. </br>Your feedback will help us to improve contribution process.
